File: python/game.py
import json
import os
import socket

from enum import Enum
from time import sleep

from utils import Position3D, Position2D, Direction


class GameInstruction(Enum):
    WALK = 1
    JUMP = 2
    ROTATE_LEFT = 3
    ROTATE_RIGHT = 4
    ROTATE_BACK = 5


class GameEvent:

    def __init__(self, instruction: GameInstruction, status: bool, message: str, position: Position3D):
        self.instruction = instruction
        self.status = status
        self.message = message
        self.position = position

    def __repr__(self):
        return f'{self.instruction.name}:{self.status.__str__().lower()}:{self.message}:{self.position.x},{self.position.y},{self.position.z}'

    def __str__(self):
        return f'GameEvent(instruction={self.instruction.name}, status={self.status}, message={self.message}, position={self.position})'


class GameRendererConnector:

    def __init__(self):
        self.socket = socket.socket()
        self.game_renderer_host = os.getenv('GAME_RENDERER_HOST', 'localhost')
        self.game_renderer_port = 7777
        self.__connect_to_renderer()

    def __connect_to_renderer(self):
        retries = 0
        while retries < 10:
            try:
                self.socket.connect((self.game_renderer_host, self.game_renderer_port))
                return
            except ConnectionRefusedError:
                retries += 1
                sleep(0.1)
        raise ConnectionRefusedError("Could not connect to game renderer")

    def send_instruction(self, instruction: GameInstruction) -> GameEvent:
        self.socket.send(instruction.name.encode('utf-8'))
        response = self.socket.recv(1024).decode('utf-8')

        response_json = json.loads(response)
        required_fields = ['status', 'message', 'position']

        if not all(field in response_json for field in required_fields):
            raise ValueError("Invalid response from game renderer")

        status = response_json['status']
        message = response_json['message']
        position = Position3D(response_json['position']['x'], response_json['position']['y'], response_json['position']['z'])

        return GameEvent(instruction, status, message, position)

    def obtain_world_info_data(self) -> dict:

        # Request world info size to know the size of the buffer to receive
        self.socket.send('world_info_size'.encode('utf-8'))
        world_info_size = int(self.socket.recv(16).decode('utf-8'))

        # Request world info
        self.socket.send('world_info'.encode('utf-8'))
        response = self.socket.recv(world_info_size).decode('utf-8')

        return json.loads(response)

    def obtain_character_status_data(self) -> dict:

        # Request character status size to know the size of the buffer to receive
        self.socket.send('character_status'.encode('utf-8'))
        response = self.socket.recv(256).decode('utf-8')

        return json.loads(response)


class GameCharacter:

    def __init__(self, game_renderer_connector: GameRendererConnector):
        self.game_renderer_connector = game_renderer_connector
        self.__events: list[GameEvent] = []
        self.current_position = Position3D(0, 0, 0)
        self.next_position = Position2D(0, 0)
        self.current_direction = Direction.UNKNOWN

    def walk(self) -> bool:
        game_event = self.game_renderer_connector.send_instruction(GameInstruction.WALK)
        self.__events.append(game_event)
        return game_event.status

    def turn_backward(self) -> bool:
        game_event = self.game_renderer_connector.send_instruction(GameInstruction.ROTATE_BACK)
        self.__events.append(game_event)
        return game_event.status

    def turn_left(self) -> bool:
        game_event = self.game_renderer_connector.send_instruction(GameInstruction.ROTATE_LEFT)
        self.__events.append(game_event)
        return game_event.status

    def turn_right(self) -> bool:
        game_event = self.game_renderer_connector.send_instruction(GameInstruction.ROTATE_RIGHT)
        self.__events.append(game_event)
        return game_event.status

    def jump(self) -> bool:
        game_event = self.game_renderer_connector.send_instruction(GameInstruction.JUMP)
        self.__events.append(game_event)
        return game_event.status

    def update_character_status(self):
        json_data = self.game_renderer_connector.obtain_character_status_data()
        character_data = json_data['character']
        self.current_position = Position3D(character_data['current_position']['x'], character_data['current_position']['y'],
                                           character_data['current_position']['z'])
        self.next_position = Position2D(character_data['next_position']['x'], character_data['next_position']['z'])
        self.current_direction = Direction(character_data['direction'])

    def apply(self):
        with open('/gamification/game_events.txt', 'w') as log_file:
            for instruction in self.__events:
                log_file.write(f'{instruction.__repr__()}\n')


class GameWorld:

    def __init__(self, game_renderer_connector: GameRendererConnector):
        self.game_renderer_connector = game_renderer_connector
        self.height_map: dict[Position2D, float] = {}

    def load_world(self):
        json_data = self.game_renderer_connector.obtain_world_info_data()
        world_data = json_data['world']
        blocks_data = world_data['blocks']

        for block in blocks_data:
            position_x = block[0]
            position_z = block[1]
            height = block[2]
            self.height_map[Position2D(position_x, position_z)] = height

    def get_height_at(self, position: Position2D) -> float:
        return self.height_map.get(position)

File: javascript/game.js
import net from 'net';
import fs from 'fs';
import {Direction} from "./utils.js";

export const GameInstruction = Object.freeze({
    WALK: Symbol("WALK"),
    JUMP: Symbol("JUMP"),
    ROTATE_LEFT: Symbol("ROTATE_LEFT"),
    ROTATE_RIGHT: Symbol("ROTATE_RIGHT"),
    ROTATE_BACK: Symbol("ROTATE_BACK"),
})

export class GameEvent {

    #instruction;
    #status;
    #message;
    #position;

    constructor(instruction, status, message, position) {
        this.#instruction = instruction;
        this.#status = status;
        this.#message = message;
        this.#position = position;
        this._instruction = instruction;
        this._message = message;
        this._position = position;
    }

    get status() {
        return this.#status;
    }

    get message() {
        return this._message;
    }

    get position() {
        return this._position;
    }

    serialize() {
        return `${this.#instruction.description}:${this.#status}:${this.#message}:${this.#position.x},${this.#position.y},${this.#position.z}`;
    }

}

export class GameRendererConnector {

    constructor() {
        this.socket = new net.Socket();
        this.gameRendererHost = process.env.GAME_RENDERER_HOST || 'localhost';
        this.gameRendererPort = 7777;
    }

    async connectToRenderer() {
        let retries = 0;

        await new Promise((resolve, reject) => {
            const attemptConnection = () => {
                const onConnect = () => {
                    resolve();
                };

                // Error handling
                const onError = (err) => {
                    if (err.code === 'ECONNREFUSED' && retries < 15) {
                        retries++;
                        // Remove the listener as it will never be called
                        this.socket.removeListener('connect', onConnect);
                        // Retry after a delay
                        setTimeout(attemptConnection, 200);
                    } else {
                        // Clean up listeners
                        this.socket.removeListener('connect', onConnect);
                        this.socket.removeListener('error', onError);
                        reject(new Error("Could not connect to game renderer: " + err.message));
                    }
                };

                // Attempt to connect
                this.socket.connect(this.gameRendererPort, this.gameRendererHost, onConnect);
                // Add error listener
                this.socket.once('error', onError);
            };

            attemptConnection();
        });
    }

    async sendInstruction(instruction) {
        // Send the instruction to the server
        this.socket.write(instruction.description);

        // Parse the response
        let response = await receiveData(this.socket);
        response = JSON.parse(response.toString());

        // Validate the response fields
        const requiredFields = ['status', 'message', 'position'];
        if (!requiredFields.every(field => field in response)) {
            throw new Error("Invalid response from game renderer");
        }

        const {status, message, position} = response;
        return new GameEvent(instruction, status, message, position);
    }

    async obtainWorldInfoData() {
        // Request world size
        this.socket.write('world_info_size');
        // Receive data from the server
        const worldInfoSize = await receiveData(this.socket);

        // Request world data
        this.socket.write('world_info');
        const worldInfoBuffer = await receiveData(this.socket);
        let worldInfo = worldInfoBuffer.toString();

        // Keep receiving data until the full world info is received
        if (worldInfo.length < worldInfoSize) {
            const additionalData = await receiveData(this.socket);
            worldInfo += additionalData.toString();
        }

        return JSON.parse(worldInfo);
    }

    async obtainCharacterStatusData() {
        // Request character status
        this.socket.write('character_status');

        // Parse the response
        const response = await receiveData(this.socket);
        return JSON.parse(response.toString());
    }

}

export class GameCharacter {

    #currentPosition;
    #nextPosition;
    #currentDirection;

    constructor(gameRendererConnector) {
        this.gameRendererConnector = gameRendererConnector;
        this.events = [];
        this.#currentPosition = {x: 0, y: 0, z: 0};
        this.#nextPosition = {x: 0, z: 0};
        this.#currentDirection = Direction.UNKNOWN;
    }

    async walk() {
        const gameEvent = await this.gameRendererConnector.sendInstruction(GameInstruction.WALK);
        this.events.push(gameEvent);
        return gameEvent.status;
    }

    async turnBackward() {
        const gameEvent = await this.gameRendererConnector.sendInstruction(GameInstruction.ROTATE_BACK);
        this.events.push(gameEvent);
        return gameEvent.status;
    }

    async turnLeft() {
        const gameEvent = await this.gameRendererConnector.sendInstruction(GameInstruction.ROTATE_LEFT);
        this.events.push(gameEvent);
        return gameEvent.status;
    }

    async turnRight() {
        const gameEvent = await this.gameRendererConnector.sendInstruction(GameInstruction.ROTATE_RIGHT);
        this.events.push(gameEvent);
        return gameEvent.status;
    }

    async jump() {
        const gameEvent = await this.gameRendererConnector.sendInstruction(GameInstruction.JUMP);
        this.events.push(gameEvent);
        return gameEvent.status;
    }

    async updateCharacterStatus() {
        const jsonData = await this.gameRendererConnector.obtainCharacterStatusData();

        this.#currentPosition = jsonData.character.current_position;
        this.#nextPosition = jsonData.character.next_position;
        this.#currentDirection = jsonData.character.direction;
    }

    apply() {
        const logFilePath = './gamification/game_events.txt';
        fs.writeFileSync(
            logFilePath,
            this.events.map(event => event.serialize()).join('\n'),
            'utf-8'
        );
    }

    get currentPosition() {
        return this.#currentPosition;
    }

    get nextPosition() {
        return this.#nextPosition;
    }

    get currentDirection() {
        return this.#currentDirection;
    }
}

export class GameWorld {

    #gameRendererConnector;
    #heightMap;

    constructor(gameRendererConnector) {
        this.#gameRendererConnector = gameRendererConnector;
        this.#heightMap = new Map(); // Using Map to replicate the behavior of a Python dictionary
    }

    async loadWorld() {
        const jsonData = await this.#gameRendererConnector.obtainWorldInfoData();
        const worldData = jsonData.world;
        const blocksData = worldData.blocks;
        const forestData = worldData.forest;
        const meshesData = worldData.extra_meshes;
        const allData = [...blocksData, ...forestData, ...meshesData];

        for (const object of allData) {
            const [positionX, positionZ, height] = object;
            const positionKey = JSON.stringify({x: positionX, z: positionZ});
            if (this.#heightMap.has(positionKey)) {
                this.#heightMap.set(positionKey, Math.max(this.#heightMap.get(positionKey), height));
            } else {
                this.#heightMap.set(positionKey, height);
            }
        }
    }

    getHeightAt(position) {
        const positionKey = JSON.stringify(position); // Convert position object to string to use as key
        return this.#heightMap.get(positionKey) || null; // Return null if no value is found
    }

}

async function receiveData(socket) {
    return new Promise((resolve, reject) => {
        // Listen for data
        const onData = (data) => {
            socket.removeListener('error', onError);
            resolve(data);
        };

        const onError = (err) => {
            socket.removeListener('data', onData);
            reject(new Error(`Socket error: ${err.message}`));
        };

        // Listen for data
        socket.once('data', onData);

        // Handle errors on the socket
        socket.once('error', onError);

        // Handle timeout if no response is received
        setTimeout(() => {
            socket.removeListener('data', onData);
            socket.removeListener('error', onError);
            reject(new Error("Response timed out"));
        }, 5000);
    });
}